src/hal/src/pso/descriptor.rs

```diff
@@ -67,6 +67,91 @@ pub enum DescriptorType {
     InputAttachment = 10,
 }
 
+impl std::convert::From<DescriptorType> for RawDescriptorType {
+    fn from(ty: DescriptorType) -> Self {
+        match ty {
+            DescriptorType::Sampler => RawDescriptorType::Sampler,
+            DescriptorType::CombinedImageSampler => RawDescriptorType::CombinedImageSampler,
+            DescriptorType::Image { ty } => match ty {
+                ImageDescriptorType::Storage => RawDescriptorType::StorageImage,
+                ImageDescriptorType::Sampled => RawDescriptorType::SampledImage,
+            }
+            DescriptorType::Buffer { access, format } => match access {
+                BufferDescriptorAccess::Storage => match format {
+                    BufferDescriptorFormat::Dynamic => RawDescriptorType::StorageBufferDynamic,
+                    BufferDescriptorFormat::Structured => RawDescriptorType::StorageBuffer,
+                    BufferDescriptorFormat::Texel => RawDescriptorType::StorageTexelBuffer,
+                },
+                BufferDescriptorAccess::Uniform => match format {
+                    BufferDescriptorFormat::Dynamic => RawDescriptorType::UniformBufferDynamic,
+                    BufferDescriptorFormat::Structured => RawDescriptorType::UniformBuffer,
+                    BufferDescriptorFormat::Texel => RawDescriptorType::UniformTexelBuffer,
+                },
+            }
+            DescriptorType::InputAttachment => RawDescriptorType::InputAttachment,
+        }
+    }
+}
+
+/// Access type of a buffer.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum BufferDescriptorAccess {
+    /// Storage buffers allow load, store, and atomic operations.
+    Storage,
```

Comment:
how do you feel about extending this to:
```rust
Storage { read_only: true }
```
I know this is a bit out of scope, but that's something WebGPU has and would be good to take advantage of on the gfx-rs level as well.
And since your PR reworks all the relevant code, it would be convenient to do it here. To clarify, there is no expectation that this boolean would affect anything today, you can just ignore it in the backends.